File: problems/Python/20. Valid Parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stk = []
        cmap = {")": "(", "}": "{", "]": "["}
        
        for c in s:
            if c in cmap:
                if not stk or stk.pop() != cmap[c]:
                    return False
            else:
                stk.append(c)
                
        return not stk

File: problems/Java/20. Valid Parentheses.java
class Solution {
    public boolean isValid(String s) {
        Deque<Character> stk = new ArrayDeque<>();
        Map<Character, Character> cmap = Map.of(')', '(', '}', '{', ']', '[');
        
        for (char c : s.toCharArray()) {
            if (cmap.containsKey(c)) {
                if (stk.isEmpty() || stk.pop() != cmap.get(c)) {
                    return false;
                }
            } else {
                stk.push(c);
            }
        }
        return stk.isEmpty();
    }
}